### ISSUE TITLE
Structs doesn't work for Arch Linux ARM

### ISSUE THREAD

**Issue by davispuh** [ffi/ffi#388]
I haven't yet really investigated this much and don't really know what's causing it, but basically, either something is trashing this struct or data isn't written to it.

Here's some C function which accepts struct as parameter.
```c
typedef enum {
  EV1,
  EV2
} myenum;

typedef struct {
  myenum s1;
  unsigned int s2;
} my_struct;

int fn(void *p1, void **p2, char const *str, my_struct mstruct, unsigned int nb, void *p3, void *p4, void *p5)
{
  int i;
  printf("%p\n", mstruct);   /*  display address to struct */
  unsigned char *p = (unsigned char *)&mstruct; /*  dump this struct */
  for (i=0; i < 8; ++i)
    printf("%02x", p[i]);
  return 0;
}
```

In ruby pass this struct with data.
```ruby
enum :MyEnum, [:EV1,
               :EV2 ]
class MyStruct < FFI::Struct
        layout({
                :s1      => :MyEnum,
                :s2      => :uint
        })
end

attach_function :fn, :fn, [ :pointer, :pointer, :string, MyStruct , :int, :pointer, :pointer, :pointer ], :uint

m_struct = MyStruct.new
m_struct[:s1] = EV1
m_struct[:s2] = 500

p = m_struct.to_ptr
puts p.inspect  # to see pointer's address
puts p.get_bytes(0, 8).unpack('H*').first  # dump struct in ruby

fn(p1, p2, "", m_struct, 0, p3, p4, p5)
```

---

On Windows 8 x64 it works fine
```
#<FFI::MemoryPointer address=0x00000002f39380 size=8>
00000000f4010000  // this is struct's value from Ruby's side
000000000023E630  // address to struct from C (I wonder why it's different than Ruby's)
00000000f4010000  // struct's value from C
```

On Arch Linux ARM (armv7l, 32-bit, Little Endian)
```
#<FFI::MemoryPointer address=0x1cf83e8 size=8>
00000000f4010000  // this is struct's value from Ruby's side
0x1cf83e8         // address to struct from C
10158d0064000000  // struct's value from C
```
for ARM that enum's value changes every time, but number is always 0x64 (100), but should be 500.

**Comment by davispuh** [comment]
I'm now looking on how to fix this, but it's not easy to understand FFI code...

so basically when `attach_function` is called with `MyStruct` (`class MyStruct < FFI::Struct`)
```ruby
attach_function :fn, :fn, [ MyStruct ], :uint
```
it should be passed by Struct's value

but when types are looked up in ([ffi/library.rb#find_type](/ffi/ffi/blob/master/lib/ffi/library.rb#L509))
```ruby
def find_type(t)
  ...
  elsif t.is_a?(Class) && t < Struct
    Type::POINTER
  elsif ...

  end || FFI.find_type(t)
end
```

it uses `Type::POINTER` type and I'm not sure if here should be separate struct type or if later it can be determined that need to pass FFI_TYPE_STRUCT ti libffi instead of FFI_TYPE_POINTER

**Comment by davispuh** [comment]
so turns out it might not be a bug but only such design that `MyStruct` is alias to `MyStruct.ptr` and so it obviously passed it as pointer. This whole time I thought that it's like in C
```c
void fn(mstruct *s1, my_struct s2);
```
would be
```ruby
attach_function :fn, :fn, [ MStruct.ptr, MyStruct ], :void
```

but actually it should be
```ruby
attach_function :fn, :fn, [ MStruct1, MyStruct.val ], :void
```
maybe it's worth changing it? so that `MyStruct` would be alias to `MyStruct.val` instead of `MyStruct.ptr` because that's more similar to how it is in C, only it looks like some tests have written expecting it to be by pointer, thus current behavior.
Changing it would be trivial
```ruby
def find_type(t)
  ...
  elsif t.is_a?(Class) && t < Struct
    t.val
  elsif ...

  end || FFI.find_type(t)
end
```


anyway, now I've changed my function definitions to use `MyStruct.val` instead and everything works correctly :)

It was this hard to find my mistake because on Windows x86/x86_64, there's no difference between `.ptr` and `.val`, structs are always passed as pointer so both ways work, but on ARM need to actually specify `.val` to pass struct by value. Also if you mix them like I did, you won't get any errros or crashes, simply target library will receive wrong data and that's why finding this wasn't easy at all.
